### PR TITLE
use enum_dispatch

### PR DESCRIPTION
**Pull request by yuunlimm**
### Description

we were initially considering using `dyn trait` to address the concerns around having too many match statements. however, since dyn trait has a performance concern, we decided to go with `enum_dispatch`, which is 10x performant

### CODE REVIEW

**Comment by yuunlimm** [comment]
* **#615** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-indexer-processors/615?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-indexer-processors/615?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by larry-aptos** [comment]
do you consider/evaluate enum_dispatch? which do you prefer(dispatch vs downcast)?

**Comment by yuunlimm** [comment]
> do you consider/evaluate enum_dispatch? which do you prefer(dispatch vs downcast)?

yeah I think enum_dispatch also works since we know there will be only fixed number of variants. and i assume enum_diospatch will be more performant than downcast?

File: rust/sdk-processor/src/parquet_processors/mod.rs [hunk 1]
```diff
@@ -56,84 +86,136 @@ pub enum ParquetTypeEnum {
...
+    ) -> anyhow::Result<()>;
+
+    fn clone_box(&self) -> Box<dyn ParquetTypeTrait>;
+}
+
+impl_downcast!(ParquetTypeTrait);
+
+/// Struct for handling Parquet types dynamically.
+#[derive(Debug)]
+pub struct ParquetTypeStructs {
```
Comment (rtso) [inline rust/sdk-processor/src/parquet_processors/mod.rs:111]: Is this struct `ParquetTypeStructs` needed if it only contains 1 field: `data`? Maybe we can use `Box<dyn ParquetTypeTrait>` directly.

File: rust/sdk-processor/src/steps/common/gcs_uploader.rs [hunk 2]
```diff
@@ -100,6 +61,10 @@ pub fn create_new_writer(schema: Arc<Type>) -> anyhow::Result<SerializedFileWrit
     SerializedFileWriter::new(Vec::new(), schema, props_arc).context("Failed to create new writer")
 }
 
+fn determine_parquet_type(buffer: &ParquetTypeStructs) -> ParquetTypeEnum {
```
Comment (larry-aptos) [inline rust/sdk-processor/src/steps/common/gcs_uploader.rs:64]: if you downcast and need the type still, I assume dispatch is better?

Comment (yuunlimm) [reply rust/sdk-processor/src/steps/common/gcs_uploader.rs:64]: agreed, I think enum dispatch is the better choice here

**Review by dermanyang — APPROVED** [review]
spoke offline; using enum_dispatch here for better performance and worth the extensibility trade-off. This crate also gives some nice code dedupe. See [enum_dispatch docs](https://docs.rs/enum_dispatch/0.3.13/enum_dispatch/index.html) for more info. 

@yuunlimm can you update the PR desc please?

File: rust/Cargo.toml [hunk 3]
```diff
@@ -146,3 +146,4 @@ serde_canonical_json = "1.0.0"
 allocative = "0.3.3"
 allocative_derive = "0.3.3"
 mockall = "0.12.1"
+downcast-rs = "1.2.1"
```
Comment (dermanyang) [inline rust/Cargo.toml:149]: leftover

**Pull request merged**